rpcs3/Emu/Cell/Modules/cellAtracXdec.cpp

```diff
@@ -0,0 +1,854 @@
+			break;
+		}
+		default:
+			fmt::throw_exception("Invalid command");
+		}
+	}
+}
+
+template <AtracXdecCmdType type>
+error_code AtracXdecContext::send_command(ppu_thread& ppu, auto&&... args)
+{
+	ppu.state += cpu_flag::wait;
+
+	{
+		std::lock_guard lock{queue_mutex};
+
+		if constexpr (type == AtracXdecCmdType::close)
+		{
+			// Close command is only sent if the queue is empty on LLE
+			if (!cmd_queue.empty())
+			{
+				return {};
+			}
+		}
+
+		if (cmd_queue.full())
+		{
+			return CELL_ADEC_ERROR_ATX_BUSY;
+		}
+
+		cmd_queue.emplace(std::forward<AtracXdecCmdType>(type), std::forward<decltype(args)>(args)...);
+	}
+
+	queue_not_empty.notify_one();
+
+	return CELL_OK;
+}
+
+void atracXdecEntry(ppu_thread& ppu, vm::ptr<AtracXdecContext> atxdec)
+{
+	atxdec->decoder.alloc_avcodec();
+
+	atxdec->exec(ppu);
+
+	atxdec->decoder.free_avcodec();
+
+	if (thread_ctrl::state() == thread_state::aborting)
+	{
+		// For savestates, restore argument
+		idm::get<named_thread<ppu_thread>>(static_cast<u32>(atxdec->thread_id))->cmd_list
+		({
+			{ ppu_cmd::set_args, 1 }, static_cast<u64>(atxdec.addr()),
+		});
+
+		ppu.state += cpu_flag::again;
+		return;
+	}
+
+	ppu_execute<&sys_ppu_thread_exit>(ppu, CELL_OK);
+}
+
+template <u32 nch_in>
+error_code _CellAdecCoreOpGetMemSize_atracx(vm::ptr<CellAdecAttr> attr)
+{
+	cellAtracXdec.notice("_CellAdecCoreOpGetMemSize_atracx<nch_in=%d>(attr=*0x%x)", nch_in, attr);
+
+	ensure(!!attr); // Not checked on LLE
+
+	constexpr u32 mem_size =
+		sizeof(AtracXdecContext) + 0x7f
+		+ ATXDEC_SPURS_STRUCTS_SIZE + 0x1d8
+		+ atracXdecGetSpursMemSize(nch_in)
+		+ ATXDEC_SAMPLES_PER_FRAME * sizeof(f32) * nch_in;
+
+	attr->workMemSize = utils::align(mem_size, 0x80);
+
+	return CELL_OK;
+}
+
+error_code _CellAdecCoreOpOpenExt_atracx(ppu_thread& ppu, vm::ptr<AtracXdecContext> handle, vm::ptr<AdecNotifyAuDone> notifyAuDone, vm::ptr<void> notifyAuDoneArg, vm::ptr<AdecNotifyPcmOut> notifyPcmOut, vm::ptr<void> notifyPcmOutArg,
+	vm::ptr<AdecNotifyError> notifyError, vm::ptr<void> notifyErrorArg, vm::ptr<AdecNotifySeqDone> notifySeqDone, vm::ptr<void> notifySeqDoneArg, vm::cptr<CellAdecResource> res, vm::cptr<CellAdecResourceSpurs> spursRes)
+{
+	std::unique_lock savestate_lock{g_fxo->get<hle_locks_t>(), std::try_to_lock};
+
+	if (!savestate_lock.owns_lock())
+	{
+		ppu.state += cpu_flag::again;
+		return {};
+	}
+
+	cellAtracXdec.notice("_CellAdecCoreOpOpenExt_atracx(handle=*0x%x, notifyAuDone=*0x%x, notifyAuDoneArg=*0x%x, notifyPcmOut=*0x%x, notifyPcmOutArg=*0x%x, notifyError=*0x%x, notifyErrorArg=*0x%x, notifySeqDone=*0x%x, notifySeqDoneArg=*0x%x, res=*0x%x, spursRes=*0x%x)",
+		handle, notifyAuDone, notifyAuDoneArg, notifyPcmOut, notifyPcmOutArg, notifyError, notifyErrorArg, notifySeqDone, notifySeqDoneArg, res, spursRes);
+
+	ensure(!!handle && !!res); // Not checked on LLE
+	ensure(handle.aligned(0x80)); // LLE cellAdec aligns the address to 128 bytes
```

Comment:
Then align it down manually.

Reply:
I think my comment there was a bit confusing. This function on LLE does not align the address manually or check the alignment. I've rewritten it to make it clearer.